fathom/src/core/pretty.rs

```diff
@@ -0,0 +1,470 @@
+}
+
+const INDENT: isize = 4;
+
+pub struct Context<'interner, 'arena> {
+    interner: &'interner RefCell<StringInterner>,
+    scope: &'arena Scope<'arena>,
+}
+
+impl<'interner, 'arena> Context<'interner, 'arena> {
+    pub fn new(
+        interner: &'interner RefCell<StringInterner>,
+        scope: &'arena Scope<'arena>,
+    ) -> Context<'interner, 'arena> {
+        Context { interner, scope }
+    }
+
+    fn string_id(&'arena self, name: StringId) -> DocBuilder<'arena, Self> {
+        match self.interner.borrow().resolve(name) {
+            Some(name) => self.text(name.to_owned()),
+            None => self.text("#error"),
+        }
+    }
+
+    pub fn module(&'arena self, module: &Module<'arena>) -> DocBuilder<'arena, Self> {
+        self.intersperse(
+            module.items.iter().map(|item| self.item(item)),
+            self.hardline(),
+        )
+    }
+
+    fn item(&'arena self, item: &Item<'arena>) -> DocBuilder<'arena, Self> {
+        match item {
+            Item::Def {
+                label,
+                r#type,
+                expr,
+            } => self
+                .concat([
+                    self.text("def"),
+                    self.space(),
+                    self.ann_pattern(Prec::Top, Some(*label), r#type),
+                    self.space(),
+                    self.text("="),
+                    self.softline(),
+                    self.term_prec(Prec::Group, expr),
+                    self.text(";"),
+                    self.hardline(),
+                ])
+                .group(),
+        }
+    }
+
+    fn pattern(&'arena self, pattern: Option<StringId>) -> DocBuilder<'arena, Self> {
+        match pattern {
+            Some(name) => self.string_id(name),
+            None => self.text("_"),
+        }
+    }
+
+    fn ann_pattern(
+        &'arena self,
+        prec: Prec,
+        pattern: Option<StringId>,
+        r#type: &Term<'arena>,
+    ) -> DocBuilder<'arena, Self> {
+        self.paren(
+            prec > Prec::Top,
+            self.concat([
+                self.concat([self.pattern(pattern), self.space(), self.text(":")])
+                    .group(),
+                self.softline(),
+                self.term_prec(Prec::Top, r#type),
+            ]),
+        )
+    }
+
+    pub fn term(&'arena self, term: &Term<'arena>) -> DocBuilder<'arena, Self> {
+        self.term_prec(Prec::Top, term)
+    }
+
+    fn term_prec(&'arena self, prec: Prec, term: &Term<'arena>) -> DocBuilder<'arena, Self> {
+        // FIXME: indentation and grouping
+
+        match term {
+            Term::ItemVar(_, level) => self.text(format!("ItemVar({:?})", level)),
+            Term::LocalVar(_, index) => self.text(format!("LocalVar({:?})", index)),
+            Term::MetaVar(_, index) => self.text(format!("MetaVar({:?})", index)),
+            Term::InsertedMeta(_, level, info) => {
+                self.text(format!("InsertedMeta({:?}, {:?})", level, info))
+            }
```

Comment:
Not a big issue, but perhaps we could do something a bit more compact for variables? Like `item#n`, `local#n`... (not sure of the naming or the syntax there). Another thought is that we could also add an environment of binder names for items, locals, and metas to make things more readable. Eg. `name#n` (still showing the level/index for debugging purposes). Not a massive issue though, just a thought!

Reply:
If the core pretty printer does include names, I would like a command line option to disable human names and only print indices: in my experience scoping/naming errors are very common. A pretty printer outputting incorrect names would be worse than useless when debugging!

Reply:
Yeah that makes sense!